winch/codegen/src/trampoline.rs

```diff
@@ -178,23 +164,196 @@ where
+                    self.masm.load(param_slot, self.scratch_reg, size);
+                    self.masm.store(self.scratch_reg.into(), store_addr, size);
+                }
+            }
+        });
+
+        self.masm.call(CalleeKind::Direct(callee_index));
+        self.masm.free_stack(total_arg_stack_space);
+        self.epilogue_with_callee_saved_restore(spill_size);
+
+        Ok(())
+    }
+
+    pub fn emit_wasm_to_native(&mut self, ty: &FuncType) -> Result<()> {
+        let mut params = Self::callee_and_caller_vmctx();
+        params.extend_from_slice(ty.params());
+
+        let callee_vmctx_size: OperandSize = params[0].into();
+
+        let func_ty = FuncType::new(params, ty.results().to_owned());
+        let wasm_sig = self.wasm_sig(&func_ty);
+        let native_sig = self.native_sig(ty);
+
+        self.prologue();
+        let (wasm_sig_stack_offsets, spill_size) = self.spill(&wasm_sig.params);
+        let total_arg_stack_space = self.reserve_stack(&wasm_sig);
+
+        // TODO Ensure save fp and pc for fast stack walking.
+        let callee_vmctx_offset = wasm_sig_stack_offsets[0];
+        let callee_offset = self.func_env.vmnative_call_host_func_native_call();
+
+        // Load the callee vm context to the scratch register.
+        let addr = self.masm.address_from_sp(callee_vmctx_offset);
+        self.masm
+            .load(addr, self.alloc_scratch_reg, callee_vmctx_size);
+
+        // Load the callee address into the other scratch register.
+        let addr = self
+            .masm
+            .address_at_reg(self.alloc_scratch_reg, callee_offset.into());
+        self.masm.load(addr, self.scratch_reg, OperandSize::S64);
+
+        // Assign the arguments.
+        native_sig.params.iter().enumerate().for_each(|(i, param)| {
+            let wasm_param = &wasm_sig.params[i];
+            let slot = match wasm_param {
+                ABIArg::Reg { .. } => self.masm.address_from_sp(wasm_sig_stack_offsets[i]),
+                ABIArg::Stack { offset, .. } => self.masm.address_at_reg(
+                    <A as ABI>::fp_reg(),
+                    *offset + (self.abi.arg_base_offset() as u32),
+                ),
+            };
+
+            match param {
+                ABIArg::Reg { reg, ty } => self.masm.load(slot, *reg, (*ty).into()),
+                ABIArg::Stack { offset, ty } => {
+                    let size = (*ty).into();
+                    let store_addr = self.masm.address_at_sp(*offset);
+                    self.masm.load(slot, self.scratch_reg, size);
+                    self.masm.store(self.scratch_reg.into(), store_addr, size);
+                }
+            }
+        });
+
+        self.masm.call(CalleeKind::Indirect(self.scratch_reg));
+        self.masm.free_stack(total_arg_stack_space);
+        self.epilogue(spill_size);
+
+        Ok(())
+    }
+
+    fn callee_and_caller_vmctx() -> Vec<ValType> {
+        vec![ValType::I64, ValType::I64]
```

Comment:
Like above, technically this `I64` should conditionally be `I32` for 32-bit hosts, right? Not that we're going to support those any time soon, but might be good to perhaps abstract ahead of time in case anyone in the future comes to add support if it's not too hard. (if it is a bit onerous I think a `// FIXME: ...` is fine too since we don't have 32-bit support anywhere on the horizon)